File: python/mcap/mcap0/chunk_builder.py
from io import BytesIO
from typing import Dict

from .data_stream import WriteDataStream
from .records import Channel, Message, MessageIndex, Schema


class ChunkBuilder:
    def __init__(self) -> None:
        self.message_end_time = 0
        self.message_indices: Dict[int, MessageIndex] = {}
        self.message_start_time = 0
        self.__buffer = BytesIO()
        self.record_writer = WriteDataStream(self.__buffer)
        self.num_messages = 0

    def data(self):
        return self.__buffer.getvalue()

    def add_channel(self, channel: Channel):
        channel.write(self.record_writer)

    def add_schema(self, schema: Schema):
        schema.write(self.record_writer)

    def add_message(self, message: Message):
        if self.message_start_time == 0:
            self.message_start_time = message.log_time
        self.message_end_time = message.log_time

        if not self.message_indices.get(message.channel_id):
            self.message_indices[message.channel_id] = MessageIndex(
                channel_id=message.channel_id, records=[]
            )
        self.message_indices[message.channel_id].records.append(
            (message.log_time, self.record_writer.count)
        )

        self.num_messages += 1
        message.write(self.record_writer)

    def reset(self):
        self.message_end_time = 0
        self.message_indices.clear()
        self.message_start_time = 0
        self.buffer = BytesIO()
        self.record_writer = WriteDataStream(self.buffer)
        self.num_messages = 0

File: typescript/mcap/src/v0/ChunkBuilder.ts
import { Mcap0RecordBuilder } from "./Mcap0RecordBuilder";
import { Channel, Message, MessageIndex, Schema } from "./types";

type ChunkBuilderOptions = {
  useMessageIndex?: boolean;
};
class ChunkBuilder {
  private recordWriter = new Mcap0RecordBuilder();
  private messageIndices: Map<number, MessageIndex> | undefined;
  private totalMessageCount = 0;

  messageStartTime = 0n;
  messageEndTime = 0n;

  constructor({ useMessageIndex = true }: ChunkBuilderOptions) {
    if (useMessageIndex) {
      this.messageIndices = new Map();
    }
  }

  get numMessages(): number {
    return this.totalMessageCount;
  }

  get buffer(): Uint8Array {
    return this.recordWriter.buffer;
  }

  get indices(): Iterable<MessageIndex> {
    if (this.messageIndices) {
      return this.messageIndices.values();
    }
    return [];
  }

  addSchema(schema: Schema): void {
    this.recordWriter.writeSchema(schema);
  }

  addChannel(info: Channel): void {
    if (this.messageIndices && !this.messageIndices.has(info.id)) {
      this.messageIndices.set(info.id, {
        channelId: info.id,
        records: [],
      });
    }
    this.recordWriter.writeChannel(info);
  }

  addMessage(message: Message): void {
    if (this.totalMessageCount === 0 || message.logTime < this.messageStartTime) {
      this.messageStartTime = message.logTime;
    }
    if (this.totalMessageCount === 0 || message.logTime > this.messageEndTime) {
      this.messageEndTime = message.logTime;
    }

    if (this.messageIndices) {
      let messageIndex = this.messageIndices.get(message.channelId);
      if (!messageIndex) {
        messageIndex = {
          channelId: message.channelId,
          records: [],
        };
        this.messageIndices.set(message.channelId, messageIndex);
      }
      messageIndex.records.push([message.logTime, BigInt(this.recordWriter.length)]);
    }

    this.totalMessageCount += 1;
    this.recordWriter.writeMessage(message);
  }

  reset(): void {
    this.messageStartTime = 0n;
    this.messageEndTime = 0n;
    this.totalMessageCount = 0;
    this.messageIndices?.clear();
  }
}

export { ChunkBuilder };
